File: credit.c
#include <stdio.h>
#include <cs50.h>
#include <math.h>

int main(void)
{

    int sums = 0;
    long creditno = 0;
    int length = 0;
    long copy;
    int firstdigit;
    int lastdigit;
    int seconddig;
    long copy2;


    do
    {
        creditno = get_long("Number: ");
        copy = creditno;
        copy2 = creditno;
    }

    while (creditno < 0);


    // Length of the numbers for validation
    while (copy != 0)
    {

        firstdigit = copy % 10;
        copy /= 10; // First digit of the cc number
        length++;
    }


    // Checksum of the credit Card
    for (int i = 0 ; creditno != 0; i++, creditno /= 10)
    {

        if (i % 2 == 0)
        {
            sums = sums + creditno % 10;
        }

        else
        {
            int digit = (creditno % 10) * 2;
            sums += (digit % 10) + (digit / 10);
        }
    }


    lastdigit = sums % 10;

    // Second digit of any number where length = number's length
    for (int i = 1; i < length ; i++)
    {
        seconddig = copy2 % 10;
        copy2 /= 10;
    }

    if (lastdigit == 0)
    {

        if ((length == 16 || length == 13) && firstdigit == 4)
        {
            printf("VISA\n");
        }

        else if (length == 16 && (seconddig == 1 || seconddig == 2 || seconddig == 3 || seconddig == 4 || seconddig == 5))
        {
            printf("MASTERCARD\n");
        }

        else if (length == 15 && (seconddig == 4 || seconddig == 7))
        {
            printf("AMEX\n");
        }

        else
        {
            printf("INVALID\n");
        }

    }

    else
    {
        printf("INVALID\n");
    }
}

File: credit.py
from cs50 import get_int


def main():
    # Prompts user for card number
    card = get_int("Number: ")

    # If card is valid, then check the card company
    if valid_card(card):
        card_company(card)
    # Card not valid
    else:
        print("INVALID")


def valid_card(num):
    # Calculates sum according to Luhn's Algorithm
    sum = 0
    for i, c in enumerate(reversed(str(num))):
        if i % 2 == 0:
            sum += int(c)
        else:
            for j in str(int(c) * 2):
                sum += int(j)

    # Checks if sum is divisible by 10
    if sum % 10 == 0:
        return True
    else:
        return False


def card_company(card):
    # Removes all the digits from card number except first 2
    num = int(str(card)[0:2])

    # Check for conditions of all companies
    if (num == 34 or num == 37) and len(str(card)) == 15:
        print("AMEX")
    elif num > 50 and num < 56 and len(str(card)) == 16:
        print("MASTERCARD")
    elif num >= 40 and num < 50 and (len(str(card)) == 13 or len(str(card)) == 16):
        print("VISA")
    else:
        print("INVALID")


if __name__ == "__main__":
    main()
